### DISCUSSION TITLE
using webpack in a react app

### DISCUSSION THREAD

**Discussion by arunkumar413** [webpack/webpack#16519] [General]
I'm trying to use webpack is a react application to bundle a virtual file system. I'm getting the below error. 

Here is the react app code;

```
import "./styles.css";
import webpack from "webpack";
import { path } from "path";

import { fs } from 'memfs';

export default function App() {
  // https://webpack.js.org/api/node

  let compiler = webpack({
    entry: "./src/index.js",
    output: {
      path: path.resolve(__dirname, "dist"),
      filename:'bundle.js'
    }
  });



  compiler.run(function (err, stats) {
    console.log(stats);

    if (err) {
      console.error(err.stack || err);
      if (err.details) {
        console.error(err.details);
      }
    }
    const info = stats.toJson();
    if (stats.hasErrors()) {
      console.error(info.errors);
    }
  });

  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
    </div>
  );
}

```
Error I'm getting

```
BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
        - add a fallback 'resolve.fallback: { "util": require.resolve("util/") }'
        - install 'util'
If you don't want to include a polyfill, you can use an empty module like this:
        resolve.fallback: { "util": false }

ERROR in ./node_modules/graceful-fs/graceful-fs.js 74:6-29
```

**Comment by alexander-akait** [comment]
Because you don't have `util` in browser, please use `fallback`, you can use `node-util` package as fallback for the browser usage

**Reply by arunkumar413kadari** [reply]
@alexander-akait I installed the `node-util ` package but still I get the same error message. What do you mean when you say `fallback`. Could you please elaborate.

**Reply by alexander-akait** [reply]
When you use node.js modules in the `web` target (i.e. for browser) you need polyfill them, because they don't exist in browsers, `fallback` for such purposes

**Reply by alexander-akait** [reply]
After installation you need to add `resolve.fallback: { "util": require.resolve("node-util") }'`

**Reply by arunkumar413** [reply]
@alexander-akait Thanks for the information. There could be other packages that are not available to run in browser environment. How to find the fallback/polyfill  packages for them.

**Reply by alexander-akait** [reply]
First, you should avoid using them when you write code for browser, some of them avaliable in Node.js (JS ecosystem is trying to be align with browser API) - https://nodejs.org/api/webcrypto.html, https://nodejs.org/api/webstreams.html and another, alos there are `node-*` packages to polyfill them (but there are limitations for some methods/functions)

**Reply by arunkumar413** [reply]
@alexander-akait Any idea tools like codesandbox and stackbllitz are able to bundle the code in the browser environment?

**Reply by alexander-akait** [reply]
Not sure, you can check it
